File: reverse_linked_list_ii.py
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None
class Solution:
    # @param {ListNode} head
    # @param {integer} m
    # @param {integer} n
    # @return {ListNode}
    def reverseBetween(self, head, m, n):
        pre=dummy = ListNode(-1)
        dummy.next = head
        count = 1
        while count < m:
            pre = pre.next
            head = head.next
            count += 1
        while count < n:
            p = head.next
            head.next = p.next
            p.next = pre.next
            pre.next = p
            count += 1
        return dummy.next
        
              


if __name__ == '__main__':
    sol = Solution()
    lst = [1,2,6,3,5,4]#
    #lst = [1,2]
    head=p= ListNode(-1)
    for value in lst:
        p.next = ListNode(value)
        p = p.next
    p.next = None
    p = sol.reverseBetween(head.next,2,4)
    print 'fuck'
    while p :
        print p.val
        p = p.next
    print 'das'

File: Algorithm/CPP/reverse_linked_list_ii.cpp
# include <stdio.h>
# include <iostream>
# include <vector>
# include <string>
# include <cctype>
# include <stack>
# include <algorithm>
# include<time.h>
using namespace std;


struct ListNode {
	int val;
	ListNode *next;
	ListNode(int x) : val(x), next(NULL) {}
};

class Solution {
public:
	ListNode* reverseBetween(ListNode* head, int m, int n) {
		ListNode* dummy = new ListNode(-1);
		dummy->next = head;
		auto p = dummy;
		auto q = head;
		int k = n - m;
		while (--m)
		{
			q = q->next;
			p = p->next;
		}
		
		while (k--)
		{
			auto nxt = q->next;
			q->next = nxt->next;
			nxt->next = p->next;
			p->next = nxt;
		}

		return dummy->next;
	}
};

int main()
{
	ListNode *head1 = new ListNode(1);
	ListNode *head2 = new ListNode(-2);
	auto p = head1;
	auto q = head2;
	vector<int> vec{ 2, 3, 5, 9 };//2
	for (auto x : vec)
	{
		p->next = new ListNode(x);
		p = p->next;
	}
	p->next = NULL;
	vec = { 3, 4, 7 };
	for (auto v : vec)
	{
		q->next = new ListNode(v);
		q = q->next;
	}
	q->next = NULL;
	//p = root->next;
	//while (p!=NULL)
	//{
	//	cout << p->val << endl;
	//	p = p->next;
	//}
	Solution sol;
	auto q1 = sol.reverseBetween(head1, 2,4);
	while (q1)
	{
		cout << q1->val;
		q1 = q1->next;
	}
	system("pause");
	return 0;
}
